tinygrad/tensor.py

```diff
@@ -119,21 +126,30 @@ def backward(self, allow_fill=True):
       assert self.data.shape == (1,)
       self.grad = Tensor(np.ones(self.data.shape, dtype=self.data.dtype), gpu=self.gpu)
 
-    assert(self.grad is not None)
-
-    with ProfileOp(self._ctx.__class__.__name__, [self.grad], backward=True):
-      grads = self._ctx.backward(self._ctx, self.grad.data)
-    if len(self._ctx.parents) == 1:
-      grads = [grads]
-    for t,g in zip(self._ctx.parents, grads):
-      if g is None:
-        continue
-      if g.shape != t.data.shape:
-        print("grad shape must match tensor shape in %r, %r != %r" %
-          (self._ctx, g.shape, t.data.shape))
-        assert(False)
-      t.grad = Tensor(g)
-      t.backward(False)
+    visited, nodes = set(), []
+    def deepwalk(node):
+      visited.add(self)
+      if node._ctx:
```

Comment:
is this ever not true?

Reply:
Yes, for all leaf nodes, we don't want to walk them